File: android/src/main/java/expo/modules/expoair/ExpoAirModule.kt
package expo.modules.expoair

import android.content.Context
import android.content.pm.ApplicationInfo
import android.content.pm.PackageManager
import android.preference.PreferenceManager
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record

class ShowBubbleOptions : Record {
    @Field var size: Double = 60.0
    @Field var color: String = "#007AFF"
}

class ExpoAirModule : Module() {

    private fun widgetBundleURL(): String? {
        val context = appContext.reactContext ?: return null

        // 1. Pre-built bundle in assets
        try {
            context.assets.open("widget.android.bundle").close()
            return "asset:///widget.android.bundle"
        } catch (_: Exception) {
            // Not found in assets
        }

        // 2. AndroidManifest meta-data for widget metro URL
        try {
            val appInfo = context.packageManager.getApplicationInfo(
                context.packageName,
                PackageManager.GET_META_DATA
            )
            val metroUrl = appInfo.metaData?.getString("expo.modules.expoair.WIDGET_METRO_URL")
            if (!metroUrl.isNullOrEmpty()) {
                return "$metroUrl/index.bundle?platform=android&dev=true&minify=false"
            }
        } catch (_: Exception) {}

        // 3. Fallback for local development (DEBUG only)
        if (isDebugBuild(context)) {
            return "http://10.0.2.2:8082/index.bundle?platform=android&dev=true&minify=false"
        }

        return null
    }

    private fun wireManagerEvents() {
        FloatingBubbleManager.onPress = {
            sendEvent("onPress", mapOf<String, Any>())
        }
        FloatingBubbleManager.onExpand = {
            sendEvent("onExpand", mapOf<String, Any>())
        }
        FloatingBubbleManager.onCollapse = {
            sendEvent("onCollapse", mapOf<String, Any>())
        }
        FloatingBubbleManager.onDragEnd = { x, y ->
            sendEvent("onDragEnd", mapOf("x" to x, "y" to y))
        }
    }

    override fun definition() = ModuleDefinition {
        Name("ExpoAir")

        Constant("PI") {
            Math.PI
        }

        Events("onChange", "onPress", "onExpand", "onCollapse", "onDragEnd")

        Function("hello") {
            "Hello world!"
        }

        AsyncFunction("setValueAsync") { value: String ->
            sendEvent("onChange", mapOf("value" to value))
        }

        Function("show") { options: ShowBubbleOptions ->
            wireManagerEvents()
            val activity = appContext.currentActivity ?: return@Function
            val bundleURL = widgetBundleURL()
            FloatingBubbleManager.show(
                activity = activity,
                size = options.size.toFloat(),
                color = options.color,
                bundleURL = bundleURL
            )
        }

        Function("hide") {
            FloatingBubbleManager.hide()
        }

        Function("expand") {
            FloatingBubbleManager.expand()
        }

        Function("collapse") {
            FloatingBubbleManager.collapse()
        }

        Function("getServerUrl") {
            val context = appContext.reactContext ?: return@Function ""

            // Check SharedPreferences first (may be set by CLI)
            val prefs = PreferenceManager.getDefaultSharedPreferences(context)
            val cached = prefs.getString("expo-air-server-url", null)
            if (!cached.isNullOrEmpty()) {
                return@Function cached
            }

            // Check AndroidManifest meta-data (set by plugin/CLI)
            try {
                val appInfo = context.packageManager.getApplicationInfo(
                    context.packageName,
                    PackageManager.GET_META_DATA
                )
                val serverUrl = appInfo.metaData?.getString("expo.modules.expoair.SERVER_URL")
                if (!serverUrl.isNullOrEmpty()) {
                    return@Function serverUrl
                }
            } catch (_: Exception) {}

            // Fallback for local development
            if (isDebugBuild(context)) {
                return@Function "ws://10.0.2.2:3847"
            }

            return@Function ""
        }

    }

    companion object {
        fun isDebugBuild(context: Context): Boolean {
            return (context.applicationInfo.flags and ApplicationInfo.FLAG_DEBUGGABLE) != 0
        }
    }
}

File: ios/ExpoAirModule.swift
import ExpoModulesCore

struct ShowBubbleOptions: Record {
  @Field var size: Double = 60
  @Field var color: String = "#000000"
}

public class ExpoAirModule: Module {
  private static func widgetBundleURL() -> URL? {
    // Production: use pre-built bundle from pod resources
    if let bundled = Bundle(for: ExpoAirModule.self).url(forResource: "widget", withExtension: "jsbundle") {
      return bundled
    }

    // Development: check for metro URL override via environment variable
    if let metroUrl = ProcessInfo.processInfo.environment["EXPO_AIR_METRO_URL"],
       let url = URL(string: "\(metroUrl)/index.bundle?platform=ios&dev=true&minify=false") {
      return url
    }

    // Development: check for metro URL from Info.plist (set by CLI)
    if let expoAir = Bundle.main.object(forInfoDictionaryKey: "ExpoAir") as? [String: Any],
       let metroUrl = expoAir["widgetMetroUrl"] as? String,
       !metroUrl.isEmpty,
       let url = URL(string: "\(metroUrl)/index.bundle?platform=ios&dev=true&minify=false") {
      return url
    }

    // Final fallback for local development only
    #if DEBUG
    return URL(string: "http://localhost:8082/index.bundle?platform=ios&dev=true&minify=false")
    #else
    return nil
    #endif
  }

  private func wireManagerEvents() {
    let manager = FloatingBubbleManager.shared
    manager.onPress = { [weak self] in
      self?.sendEvent("onPress", [:])
    }
    manager.onExpand = { [weak self] in
      self?.sendEvent("onExpand", [:])
    }
    manager.onCollapse = { [weak self] in
      self?.sendEvent("onCollapse", [:])
    }
    manager.onDragEnd = { [weak self] x, y in
      self?.sendEvent("onDragEnd", ["x": x, "y": y])
    }
    manager.onActionPress = { [weak self] in
      NSLog("[ExpoAirModule] onActionPress callback fired, sending event to host JS")
      self?.sendEvent("onActionPress", [:])
    }
  }

  public func definition() -> ModuleDefinition {
    Name("ExpoAir")

    Constant("PI") {
      Double.pi
    }

    Events("onChange", "onPress", "onExpand", "onCollapse", "onDragEnd", "onActionPress")

    Function("hello") {
      return "Hello world!"
    }

    AsyncFunction("setValueAsync") { (value: String) in
      self.sendEvent("onChange", [
        "value": value
      ])
    }

    Function("show") { (options: ShowBubbleOptions) in
      self.wireManagerEvents()
      FloatingBubbleManager.shared.show(
        size: CGFloat(options.size),
        color: options.color,
        bundleURL: Self.widgetBundleURL()
      )
    }

    Function("hide") {
      FloatingBubbleManager.shared.hide()
    }

    Function("expand") {
      FloatingBubbleManager.shared.expand()
    }

    Function("collapse") {
      FloatingBubbleManager.shared.collapse()
    }

    Function("setServerUrl") { (url: String) in
      FloatingBubbleManager.shared.updateServerUrl(url)
    }

    Function("reloadWidget") { (bundleUrlString: String, serverUrlString: String) in
      guard let bundleUrl = URL(string: bundleUrlString) else { return }
      FloatingBubbleManager.shared.reloadWidget(bundleURL: bundleUrl, serverUrl: serverUrlString)
    }

    Function("setAction") { (config: [String: Any]?) in
      NSLog("[ExpoAirModule] setAction called with config: %@", config?.description ?? "nil")
      FloatingBubbleManager.shared.updateAction(config)
    }

    Function("getServerUrl") { () -> String in
      // Check UserDefaults first (may be set by CLI)
      if let cached = UserDefaults.standard.string(forKey: "expo-air-server-url"), !cached.isEmpty {
        return cached
      }
      // Check Info.plist (set by plugin/CLI)
      if let expoAir = Bundle.main.object(forInfoDictionaryKey: "ExpoAir") as? [String: Any],
         let serverUrl = expoAir["serverUrl"] as? String,
         !serverUrl.isEmpty {
        return serverUrl
      }
      // Fallback for local development
      #if DEBUG
      return "ws://localhost:3847"
      #else
      return ""
      #endif
    }

  }
}
